qubes/tests/__init__.py

```diff
@@ -371,16 +374,55 @@ def __str__(self):
 
     def setUp(self):
         super().setUp()
-        self.loop = asyncio.new_event_loop()
-        asyncio.set_event_loop(self.loop)
+        self.addCleanup(self.cleanup_gc)
+
+        self.loop = asyncio.get_event_loop()
         self.addCleanup(self.cleanup_loop)
 
+    def cleanup_gc(self):
+        gc.collect()
+        leaked = [obj for obj in gc.get_objects() + gc.garbage
+            if isinstance(obj,
+                (qubes.Qubes, qubes.vm.BaseVM,
+                libvirt.virConnect, libvirt.virDomain))]
+
+        if leaked:
+            try:
+                import objgraph
+                objgraph.show_backrefs(leaked,
+                    max_depth=15, extra_info=extra_info,
+                    filename='/tmp/objgraph-{}.png'.format(self.id()))
+            except ImportError:
+                pass
+
+        assert not leaked
+
     def cleanup_loop(self):
-        # The loop, when closing, throws a warning if there is
-        # some unfinished bussiness. Let's catch that.
-        with warnings.catch_warnings():
-            warnings.simplefilter('error')
-            self.loop.close()
+        '''Check if the loop is empty'''
+        # XXX BEWARE this is touching undocumented, implementation-specific
+        # attributes of the loop. This is most certainly unsupported and likely
+        # will break when messing with: Python version, kernel family, loop
+        # implementation, a combination thereof, or other things.
+        # KEYWORDS for searching:
+        #   win32, SelectorEventLoop, ProactorEventLoop, uvloop, gevent
+
+        global libvirt_event_impl
+
+        # Check for unfinished libvirt business.
+        if libvirt_event_impl is not None:
+            self.loop.run_until_complete(libvirt_event_impl.drain())
```

Comment:
Is the patch adding `drain()` accepted upstream? AFAIR its original form got rejected.

Reply:
No, it got rejected and I haven't resubmitted, since upstream was around version bump. I have the updated series though. Would you like it to be submitted upstream, included in core-libvirt or both?

Reply:
Both. Replacing https://github.com/QubesOS/qubes-core-libvirt/pull/4